### PR TITLE
Ruby: Data flow for keyword arguments/parameters

### PR DESCRIPTION
**Pull request by hvitved**
Adds support for flow into keyword parameters from keyword arguments. Also gets rid of the `-1` and `-2` positions for modelling `self` and block parameters, respectively.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/dataflow/internal/FlowSummaryImplSpecific.qll [hunk 1]
```diff
@@ -63,12 +63,13 @@ SummaryComponent interpretComponentSpecific(string c) {
...
   or
   c = "Argument[_]" and
-  result = FlowSummary::SummaryComponent::argument(any(int i | i >= 0))
+  result = FlowSummary::SummaryComponent::argument(any(ParameterPosition pos | pos.isPositional(_)))
 }
 
 /** Gets the textual representation of a summary component in the format used for flow summaries. */
 string getComponentSpecificCsv(SummaryComponent sc) {
-  sc = TArgumentSummaryComponent(-2) and result = "BlockArgument"
+  sc = TArgumentSummaryComponent(any(ParameterPosition pos | pos.isSelf())) and
```
[the remark this answers is not in the document]
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/FlowSummaryImplSpecific.qll:71]: Indeed; good catch.

[the remark this answers is not in the document]
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/FlowSummaryImplSpecific.qll:71]: (The predicate is only intended to be used for testing, and is currently unused for Ruby.)

File: ruby/ql/lib/codeql/ruby/dataflow/internal/FlowSummaryImplSpecific.qll [hunk 2]
```diff
@@ -12,7 +12,7 @@ private import FlowSummaryImpl::Public
 private import codeql.ruby.dataflow.FlowSummary as FlowSummary
 
 /** Gets the parameter position of the instance parameter. */
-int instanceParameterPosition() { none() } // disables implicit summary flow to `self` for callbacks
+ArgumentPosition instanceParameterPosition() { none() } // disables implicit summary flow to `self` for callbacks
```
Comment (alexrford) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/FlowSummaryImplSpecific.qll:15]: I _think_ that this has the same explanation as my comment before, but it's a bit surprising to see `Gets the parameter position` in the QLDoc but `ArgumentPosition` as the result type.

**Review by alexrford — COMMENTED** [review]
The changes seem to make sense, and I'm very happy to see the special cases for `-1` and `-2` gone.

Just a couple of questions around some `ArgumentPosition` and `ParameterPosition` uses. I _think_ that these are right, but I had to stare at them for a while - so I'd like to check my understanding.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll [hunk 3]
```diff
@@ -230,6 +231,30 @@ private module Cached {
...
+  cached
+  newtype TParameterPosition =
+    TSelfParameterPosition() or
+    TBlockParameterPosition() or
+    TPositionalParameterPosition(int pos) {
+      pos = any(Parameter p).getPosition()
+      or
+      pos in [0 .. 10] // TODO: remove once `Argument[_]` summaries are replaced with `Argument[i..]`
+      or
+      FlowSummaryImplSpecific::ParsePositions::isParsedArgumentPosition(_, pos)
```
[the remark this answers is not in the document]
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:255]: Yes, that is correct.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/FlowSummaryImplSpecific.qll (hunk 2 above)
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/FlowSummaryImplSpecific.qll:15]: Yes, correct. I agree that the QL doc is not optimal.

**Review by alexrford — APPROVED** [review]

**Pull request merged**